src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs

```diff
@@ -131,44 +142,56 @@ public static void AreAlmostEqual(FileInfo expected, Rectangle expectedRect, Fil
 							continue;
 						}
 
-						var cumulativeError = Abs(expectedPixel.R - actualPixel.R)
-							+ Abs(expectedPixel.G - actualPixel.G)
-							+ Abs(expectedPixel.B - actualPixel.B)
-							+ Abs(expectedPixel.A - actualPixel.A);
-
+						var cumulativeError = GetColorDistance(expectedPixel, actualPixel);
 						if (cumulativeError > permittedPixelError)
 						{
-							Assert.Fail($"Difference between expected={expectedPixel} and actual={actualPixel} at {x},{y} exceeds permitted error of {permittedPixelError}");
+							Assert.Fail(WithContext(builder: builder => builder
+								.AppendLine($"Difference between pixels at ({x},{y}) exceeds permitted error of {permittedPixelError}")
+								.AppendLine($"expectedPixel: {ToArgbCode(expectedPixel)} {expectedPixel}")
+								.AppendLine($"actualPixel  : {ToArgbCode(actualPixel)} {actualPixel}")
+								.AppendLine($"permittedPixelError: {permittedPixelError}")
+								.AppendLine($"cumulativeError    : {cumulativeError}")
+							));
 						}
 					}
 			}
-		}
 
-		public static void AreNotEqual(FileInfo expected, FileInfo actual, IAppRect rect)
-			=> AreNotEqual(expected, actual, new Rectangle((int)rect.X, (int)rect.Y, (int)rect.Width, (int)rect.Height));
-		public static void AreNotEqual(FileInfo expected, FileInfo actual, Rectangle? rect = null)
-		{
-			rect = rect ?? new Rectangle(0, 0, int.MaxValue, int.MaxValue);
-			AreNotEqual(expected, rect.Value, actual, rect.Value);
+			string WithContext(string message = null, Action<StringBuilder> builder = null)
+			{
+				return new StringBuilder()
+					.AppendLine($"ImageAssert.AreAlmostEqual @ line {line}")
+					.AppendLine("permittedPixelError: " + permittedPixelError)
+					.AppendLine("expected: " + expected?.Name + (expectedRect == FirstQuadrant ? null : $"in {expectedRect}"))
+					.AppendLine("actual  : " + actual?.Name + (actualRect == FirstQuadrant ? null : $"in {actualRect}"))
+					.AppendLine("====================")
+					.ApplyIf(message != null, x => x.AppendLine(message))
+					.ApplyIf(builder != null, builder)
+					.ToString();
+			}
 		}
+		#endregion
+
+		#region AreNotEqual
+		public static void AreNotEqual(FileInfo expected, FileInfo actual, IAppRect rect, [CallerLineNumber] int line = 0)
+			=> AreNotEqualImpl(expected, rect.ToRectangle(), actual, rect.ToRectangle(), line);
+
+		public static void AreNotEqual(FileInfo expected, FileInfo actual, Rectangle? rect = null, [CallerLineNumber] int line = 0)
+			=> AreNotEqualImpl(expected, rect ?? FirstQuadrant, actual, rect ?? FirstQuadrant, line);
+
+		public static void AreNotEqual(FileInfo expected, IAppRect expectedRect, FileInfo actual, IAppRect actualRect, [CallerLineNumber] int line = 0)
+			=> AreNotEqualImpl(expected, expectedRect.ToRectangle(), actual, actualRect.ToRectangle(), line);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'actualRect' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5170992)